ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -254,6 +245,33 @@ public boolean updateDataNodeInfoInCluster(ReplicaId replicaId, boolean shouldEx
     return Collections.unmodifiableMap(partitionStateChangeListeners);
   }
 
+  @Override
+  public void setReplicaDisabledState(ReplicaId replicaId, boolean disable) {
+    if (!(replicaId instanceof AmbryReplica)) {
+      throw new IllegalArgumentException(
+          "HelixParticipant only works with the AmbryReplica implementation of ReplicaId");
+    }
+    synchronized (helixAdministrationLock) {
+      List<String> disabledReplicas = new ArrayList<>(getDisabledReplicas());
+      String partitionName = replicaId.getPartitionId().toPathString();
+      if (!disable && disabledReplicas.contains(partitionName)) {
+        logger.info("Removing the partition {} from disabledReplicas list", partitionName);
+        disabledReplicas.remove(partitionName);
+        setDisabledReplicas(disabledReplicas);
+      } else if (disable && !disabledReplicas.contains(partitionName)) {
+        logger.info("Adding the partition {} to disabledReplicas list", partitionName);
+        disabledReplicas.add(partitionName);
+        setDisabledReplicas(disabledReplicas);
+      }
+
+      // invoke Helix native method to enable/disable partition on local node, this will trigger subsequent state transition
+      InstanceConfig instanceConfig = getInstanceConfig();
+      String resourceNameForPartition = getResourceNameOfPartition(helixAdmin, clusterName, partitionName);
+      instanceConfig.setInstanceEnabledForPartition(resourceNameForPartition, partitionName, !disable);
```

Comment:
it looks like this method just updates the ZnRecord java object. Do you need to call HelixAdmin.setInstanceConfig like was done in `setDisabledReplicas`. Also, I was wondering if we need the additional `disabledReplicas` field or if we could just use `HELIX_DISABLED_PARTITION`? I may be missing something, but it seems like they hold the same set of information and are updated together at the same time.

Reply:
Good question. Yes, you are right, `disabledReplicas` and `HELIX_DISABLED_PARTITION` hold the same information. Let me explain a little more.
1. Why we need an extra disabled list?  `setInstanceEnabledForPartition` is Helix native method and is also used by "move replica" feature. When we decommission a replica, the first step is to call this method to disable replica. It will add `HELIX_DISABLED_PARTITION` map in InstanceConfig. There is a problem that when server restarts, it sees this map and cannot tell if the replica is disabled because of disk failure or because of replica decommission. Hence, I add this `disabledReplicas` to only hold replicas that are disabled due to disk failure and server will attempt to re-enable them.
2. Why we don't call `setInstanceConfig` again? `setInstanceEnabledForPartition` automatically updates InstanceConfig by change MapFields. Looks like it's not necessary to set InstanceConfig again. I also hesitate on calling `setInstanceConfig` as it overrides all fields (ListFields, MapFields and SimpleFields)
Hope this comment makes more sense.

Reply:
Thanks for clarifying point 1. 

For 2, I am still a little confused about how `setInstanceEnabledforPartition` ensures that the change is reflected in zookeeper. It looks like all the implementation does is call `_record.setListField(InstanceConfigProperty.HELIX_DISABLED_PARTITION.toString(), list);`

This seems to only change the local state of the java object, but I thought that a call to HelixAdmin.setInstanceConfig was needed to actually call zookeeper client. This is what is done in the implementation of `setSealedReplicas`

Reply:
yeah, you are right. I checked my previous usage of `setInstanceEnabledForPartition` in `HelixBootstrapUpgradeUtil`, it did call `setInstanceConfig` in the end. I will correct this part and probably re-arrange the order: 
 step1: calling `setInstanceEnabledForPartition`, make sure it updates the map fields (for our new version, it won't call `_record.setListField`)
 step2: updating our own disabled replica list
 step3: calling `setInstanceConfig` to make both of them take effect
What do you think?   (changing the order is to avoid `setInstanceEnabledForPartition` overrides the list field, although it should not)

Reply:
Yep, I think that order makes sense